## 4.9 Data Visualization with Python - Task Part 1

### Contents
01 Import Libaries and Customers Data

02 View Customers Dataframe

03 Rename Columns in Customers Dataframe

04 View Descriptive Statistics of Customers Dataframe

05 Check for Duplicates in Customers

06 Check for Missing Values in Customers

07 Check for Mixed Data Types in Customers

08 Merge Customers Data with Prepared Data

09 Reduce Data Type for Memory Savings

10 Export Merged Data

### 01 Import Libaries and Customers Data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
# Create path variable for main project folder
path = r'D:\JupyterProjects\06-2022 Instacart Basket Analysis'

In [3]:
# Import new customers data
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

### 02 View Customers Dataframe

In [4]:
# View top five rows
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# View dimensions of dataframe
customers.shape

(206209, 10)

In [6]:
# View summary
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


### 03 Rename Columns in Customers Dataframe

In [7]:
# Rename First Name column
customers.rename(columns = {'First Name': 'first_name'}, inplace = True)

In [8]:
customers.rename(columns = {'Surnam': 'last_name'}, inplace = True)

In [9]:
customers.rename(columns = {'Gender': 'gender'}, inplace = True)

In [10]:
customers.rename(columns = {'STATE': 'state'}, inplace = True)

In [11]:
customers.rename(columns = {'Age': 'age'}, inplace = True)

In [12]:
customers.rename(columns = {'n_dependants': 'no_of_dependents'}, inplace = True)

In [13]:
# Recheck column headings
customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no_of_dependents fam_status  income  
0                 3    married  165665  
1                 0     single   59285  
2                 2    married   99568  
3                 0     single   42049  
4                 1    married   40374

### 04 View Descriptive Statistics of Customers Dataframe

In [14]:
customers.describe()

user_id            age  no_of_dependents         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

### 05 Check for Duplicates in Customers

In [15]:
# Check for full duplicates
df_dups = customers[customers.duplicated()]

In [16]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, no_of_dependents, fam_status, income]
Index: []

No full duplicate records found.

### 06 Check for Missing Values in Customers

In [17]:
# Check for null values
customers.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependents        0
fam_status              0
income                  0
dtype: int64

In [18]:
# View missing data
df_nan = customers[customers['first_name'].isnull() == True]

In [19]:
df_nan.head(10)

user_id first_name   last_name  gender           state  age date_joined  \
53     76659        NaN     Gilbert    Male        Colorado   26    1/1/2017   
73     13738        NaN       Frost  Female       Louisiana   39    1/1/2017   
82     89996        NaN      Dawson  Female          Oregon   52    1/1/2017   
99     96166        NaN     Oconnor    Male        Oklahoma   51    1/1/2017   
105    29778        NaN      Dawson  Female            Utah   63    1/1/2017   
128     8562        NaN     Oconnor    Male            Utah   46    1/1/2017   
140   149267        NaN  Hutchinson    Male  South Carolina   20    1/1/2017   
149    82632        NaN         Orr    Male          Hawaii   61    1/1/2017   
155   172331        NaN  Williamson  Female          Alaska   27    1/1/2017   
236   182963        NaN   Nicholson  Female      New Mexico   58    1/2/2017   

     no_of_dependents fam_status  income  
53                  2    married   41709  
73                  0     single   82518  
82                  3    married  117099  
99                  1    married  155673  
105                 3    married  151819  
128                 1    married  134898  
140                 0     single   86778  
149                 1    married  118130  
155                 0     single   55047  
236                 1    married  163391

In [20]:
# Verify unique values for user_id column
df_unique = customers['user_id'].duplicated()

In [21]:
# View count 
df_unique.value_counts()

False    206209
Name: user_id, dtype: int64

There are 11,259 first names missing, but all have a unique user_id which will be the key characteristic.  Leaving as is for now.

### 07 Check for Mixed Data Types in Customers

In [22]:
# View column data types
customers.dtypes

user_id              int64
first_name          object
last_name           object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependents     int64
fam_status          object
income               int64
dtype: object

In [23]:
# Run test for mixed data types
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

first_name


In [24]:
# Change first_name column data type from object to string
customers['first_name'] = customers['first_name'].astype('str')

In [25]:
# Rerun test
for col in customers.columns.tolist():
  weird = (customers[[col]].applymap(type) != customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customers[weird]) > 0:
    print (col)

### 08 Merge Customers Data with Prepared Data

In [26]:
# Import prepared data
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'Four8Final.pkl'))

In [27]:
# View data
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered  ...    price_range_loc     busiest_day    busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regularly busy   
1          1  ...  Mid-range product  Regularly busy    Slowest days   
2          1  ...  Mid-range product  Regularly busy    Slowest days   
3          1  ...  Mid-range product      Least busy    Slowest days   
4          1  ...  Mid-range product      Least busy    Slowest days   

   busiest_period_of_day max_order  loyalty_flag mean_price spending_flag  \
0         Average orders        10  New customer   6.367188   Low spender   
1         Average orders        10  New customer   6.367188   Low spender   
2         Average orders        10  New customer   6.367188   Low spender   
3         Average orders        10  New customer   6.367188   Low spender   
4         Average orders        10  New customer   6.367188   Low spender   

  median_days_prior         frequency_flag  
0              20.5  Non-frequent customer  
1              20.5  Non-frequent customer  
2              20.5  Non-frequent customer  
3              20.5  Non-frequent customer  
4              20.5  Non-frequent customer  

[5 rows x 25 columns]

In [28]:
# Check shape
ords_prods_merge.shape

(32404859, 25)

In [29]:
# Check data type of matching field
ords_prods_merge['user_id'].dtype

dtype('int32')

In [30]:
# Change data type of user_id in customers to match data type in ords_prods_merge

In [31]:
# Verify max of user_id in customers
customers['user_id'].max()

206209

In [32]:
# Change data type of user_id to int to match data type in ords_prods_merge
customers = customers.astype({'user_id': 'int32'})

In [33]:
# Merge customers dataframe with ords_prods_merge dataframe
merged_final = ords_prods_merge.merge(customers, on = 'user_id')

In [34]:
# Check shape of new dataframe
merged_final.shape

(32404859, 34)

In [35]:
# Check head of merged dataframe
merged_final.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   
5                    19.0         False         196                  1   
6                    20.0         False         196                  1   
7                    14.0         False         196                  2   
8                     0.0         False         196                  4   
9                    30.0         False         196                  1   

   reordered  ...         frequency_flag  first_name  last_name  gender  \
0          0  ...  Non-frequent customer       Linda     Nguyen  Female   
1          1  ...  Non-frequent customer       Linda     Nguyen  Female   
2          1  ...  Non-frequent customer       Linda     Nguyen  Female   
3          1  ...  Non-frequent customer       Linda     Nguyen  Female   
4          1  ...  Non-frequent customer       Linda     Nguyen  Female   
5          1  ...  Non-frequent customer       Linda     Nguyen  Female   
6          1  ...  Non-frequent customer       Linda     Nguyen  Female   
7          1  ...  Non-frequent customer       Linda     Nguyen  Female   
8          1  ...  Non-frequent customer       Linda     Nguyen  Female   
9          1  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined no_of_dependents fam_status  income  
0  Alabama  31   2/17/2019                3    married   40423  
1  Alabama  31   2/17/2019                3    married   40423  
2  Alabama  31   2/17/2019                3    married   40423  
3  Alabama  31   2/17/2019                3    married   40423  
4  Alabama  31   2/17/2019                3    married   40423  
5  Alabama  31   2/17/2019                3    married   40423  
6  Alabama  31   2/17/2019                3    married   40423  
7  Alabama  31   2/17/2019                3    married   40423  
8  Alabama  31   2/17/2019                3    married   40423  
9  Alabama  31   2/17/2019                3    married   40423  

[10 rows x 34 columns]

### 09 Reduce Data Type for Memory Savings

In [36]:
# Check data types
merged_final.dtypes

order_id                     int32
user_id                      int32
order_number                  int8
orders_day_of_week            int8
order_hour_of_day             int8
days_since_prior_order     float16
new_customer                  bool
product_id                   int32
add_to_cart_order            int16
reordered                     int8
product_name                object
aisle_id                     int16
department_id                 int8
prices                     float16
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int16
loyalty_flag                object
mean_price                 float16
spending_flag               object
median_days_prior          float16
frequency_flag              object
first_name                  object
last_name                   object
gender                      object
state               

In [37]:
merged_final['age'].max()

81

In [38]:
merged_final = merged_final.astype({'age': 'int8'})

In [39]:
merged_final['no_of_dependents'].max()

3

In [40]:
merged_final = merged_final.astype({'no_of_dependents': 'int8'})

In [41]:
merged_final['income'].max()

593901

In [42]:
merged_final = merged_final.astype({'income': 'int32'})

In [44]:
# Recheck data types
merged_final.dtypes

order_id                     int32
user_id                      int32
order_number                  int8
orders_day_of_week            int8
order_hour_of_day             int8
days_since_prior_order     float16
new_customer                  bool
product_id                   int32
add_to_cart_order            int16
reordered                     int8
product_name                object
aisle_id                     int16
department_id                 int8
prices                     float16
_merge                    category
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int16
loyalty_flag                object
mean_price                 float16
spending_flag               object
median_days_prior          float16
frequency_flag              object
first_name                  object
last_name                   object
gender                      object
state               

### 10 Export Merged Data

In [43]:
# Export as pickle file
merged_final.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'merged_final.pkl'))